File: backend/utils/seats.py
"""
Seat utilities for bridge position calculations.

This module provides a single source of truth for all seat-related
calculations across the application. Use these functions instead of
inline seat math to ensure consistency.

Seat Mapping:
    NORTH = 0, EAST = 1, SOUTH = 2, WEST = 3

The modulo-4 clock:
    - Partner is always +2 (opposite)
    - LHO (Left Hand Opponent) is always +1 (clockwise)
    - RHO (Right Hand Opponent) is always +3 (or -1)

Usage:
    from utils.seats import partner, lho, rho, relative_position

    # Get partner of South
    partner('S')  # Returns 'N'

    # Get display name relative to user
    display_name('E', hero='S')  # Returns 'LHO'

    # Check if two seats are partners
    is_partner('N', 'S')  # Returns True
"""

from typing import Optional, Set, Literal

# === CONSTANTS ===

# Integer indices (for modulo arithmetic)
NORTH: int = 0
EAST: int = 1
SOUTH: int = 2
WEST: int = 3

# Partnerships
NS: int = 0  # North-South
EW: int = 1  # East-West

# String representations
SEATS: list[str] = ['N', 'E', 'S', 'W']
SEAT_NAMES: dict[str, str] = {'N': 'North', 'E': 'East', 'S': 'South', 'W': 'West'}
RELATIVE_NAMES: list[str] = ['You', 'LHO', 'Partner', 'RHO']

# Partnership sets
NS_SIDE: Set[str] = {'N', 'S'}
EW_SIDE: Set[str] = {'E', 'W'}

# Direct lookup maps (for performance when not using indices)
PARTNERS: dict[str, str] = {'N': 'S', 'S': 'N', 'E': 'W', 'W': 'E'}
NEXT_PLAYER: dict[str, str] = {'N': 'E', 'E': 'S', 'S': 'W', 'W': 'N'}  # LHO / clockwise
PREV_PLAYER: dict[str, str] = {'N': 'W', 'E': 'N', 'S': 'E', 'W': 'S'}  # RHO / counter-clockwise


# === TYPE ALIASES ===

SeatStr = Literal['N', 'E', 'S', 'W']
SeatFull = Literal['North', 'East', 'South', 'West']


# === NORMALIZATION ===

def normalize(position: Optional[str]) -> str:
    """
    Normalize a position to single-letter format.

    Args:
        position: Position string ('North', 'N', 'north', etc.)

    Returns:
        Single letter format ('N', 'E', 'S', 'W')

    Examples:
        normalize('North') -> 'N'
        normalize('s') -> 'S'
        normalize(None) -> 'N'
    """
    if not position:
        return 'N'  # Default

    pos = str(position).strip()

    # Already single letter
    if pos.upper() in SEATS:
        return pos.upper()

    # Full name
    full_map = {'NORTH': 'N', 'EAST': 'E', 'SOUTH': 'S', 'WEST': 'W'}
    return full_map.get(pos.upper(), 'N')


def seat_index(seat: str) -> int:
    """
    Convert seat string to index (0-3).

    Args:
        seat: Seat string ('N', 'E', 'S', 'W' or full name)

    Returns:
        Integer index (0=N, 1=E, 2=S, 3=W)
    """
    return SEATS.index(normalize(seat))


def seat_from_index(index: int) -> str:
    """
    Convert index to seat string.

    Args:
        index: Integer index (0-3, will be wrapped with modulo)

    Returns:
        Seat string ('N', 'E', 'S', 'W')
    """
    return SEATS[index % 4]


# === RELATIONSHIP FUNCTIONS ===

def partner(seat: str) -> str:
    """
    Get the partner of a seat.

    Args:
        seat: Seat string

    Returns:
        Partner's seat string

    Example:
        partner('S') -> 'N'
        partner('E') -> 'W'
    """
    s = normalize(seat)
    return PARTNERS[s]


def lho(seat: str) -> str:
    """
    Get the Left Hand Opponent (next player clockwise).

    Args:
        seat: Seat string

    Returns:
        LHO's seat string

    Example:
        lho('S') -> 'W'
        lho('N') -> 'E'
    """
    s = normalize(seat)
    return NEXT_PLAYER[s]


def rho(seat: str) -> str:
    """
    Get the Right Hand Opponent (previous player clockwise).

    Args:
        seat: Seat string

    Returns:
        RHO's seat string

    Example:
        rho('S') -> 'E'
        rho('N') -> 'W'
    """
    s = normalize(seat)
    return PREV_PLAYER[s]


def partnership(seat: str) -> int:
    """
    Get the partnership of a seat.

    Args:
        seat: Seat string

    Returns:
        Partnership index (0=NS, 1=EW)
    """
    return seat_index(seat) % 2


def partnership_str(seat: str) -> str:
    """
    Get the partnership name of a seat.

    Args:
        seat: Seat string

    Returns:
        Partnership string ('NS' or 'EW')
    """
    return 'NS' if partnership(seat) == NS else 'EW'


def is_partner(seat1: str, seat2: str) -> bool:
    """
    Check if two seats are partners.

    Args:
        seat1: First seat string
        seat2: Second seat string

    Returns:
        True if seats are partners
    """
    return partner(seat1) == normalize(seat2)


def is_opponent(seat1: str, seat2: str) -> bool:
    """
    Check if two seats are opponents.

    Args:
        seat1: First seat string
        seat2: Second seat string

    Returns:
        True if seats are opponents
    """
    return partnership(seat1) != partnership(seat2)


def same_side(seat1: str, seat2: str) -> bool:
    """
    Check if two seats are on the same side (partners).

    Args:
        seat1: First seat string
        seat2: Second seat string

    Returns:
        True if seats are on the same partnership
    """
    return partnership(seat1) == partnership(seat2)


# === RELATIVE POSITION ===

def relative_position(target: str, hero: str) -> int:
    """
    Get the relative position of target from hero's perspective.

    Uses modulo-4 clock arithmetic:
        0 = Self
        1 = LHO (Left Hand Opponent)
        2 = Partner
        3 = RHO (Right Hand Opponent)

    Args:
        target: Target seat string
        hero: Hero/user seat string

    Returns:
        Relative position index (0-3)

    Example:
        relative_position('W', 'S') -> 1 (LHO)
        relative_position('N', 'S') -> 2 (Partner)
        relative_position('E', 'S') -> 3 (RHO)
    """
    return (seat_index(target) - seat_index(hero)) % 4


def display_name(seat: str, hero: str = 'S', relative: bool = True) -> str:
    """
    Get a display name for a seat.

    Args:
        seat: Target seat string
        hero: Hero/user seat string (default 'S')
        relative: If True, return relative name ('Partner');
                  if False, return absolute name ('North')

    Returns:
        Display name string

    Example:
        display_name('N', hero='S') -> 'Partner'
        display_name('N', hero='S', relative=False) -> 'North'
        display_name('E', hero='S') -> 'RHO'
    """
    s = normalize(seat)
    if relative:
        rel = relative_position(s, hero)
        return RELATIVE_NAMES[rel]
    return SEAT_NAMES[s]


def bidder_role(bidder: str, user: str) -> str:
    """
    Get the role description for a bidder relative to the user.

    Args:
        bidder: Bidder's seat string
        user: User's seat string

    Returns:
        Role string with seat name, e.g., 'North (Partner)'

    Example:
        bidder_role('N', 'S') -> 'North (Partner)'
        bidder_role('E', 'S') -> 'East (Opponent)'
        bidder_role('S', 'S') -> 'South (You)'
    """
    b = normalize(bidder)
    u = normalize(user)

    rel = relative_position(b, u)
    seat_name = SEAT_NAMES[b]

    if rel == 0:
        return f"{seat_name} (You)"
    elif rel == 2:
        return f"{seat_name} (Partner)"
    else:
        return f"{seat_name} (Opponent)"


# === PLAY PHASE HELPERS ===

def dummy(declarer: str) -> str:
    """
    Get the dummy seat given the declarer.

    Args:
        declarer: Declarer's seat string

    Returns:
        Dummy's seat string (declarer's partner)
    """
    return partner(declarer)


def opening_leader(declarer: str) -> str:
    """
    Get the opening leader seat given the declarer.

    Args:
        declarer: Declarer's seat string

    Returns:
        Opening leader's seat string (LHO of declarer)
    """
    return lho(declarer)


def is_declaring_side(seat: str, declarer: str) -> bool:
    """
    Check if a seat is on the declaring side.

    Args:
        seat: Seat to check
        declarer: Declarer's seat string

    Returns:
        True if seat is declarer or dummy
    """
    return same_side(seat, declarer)


def is_defending_side(seat: str, declarer: str) -> bool:
    """
    Check if a seat is on the defending side.

    Args:
        seat: Seat to check
        declarer: Declarer's seat string

    Returns:
        True if seat is a defender
    """
    return not is_declaring_side(seat, declarer)


# === BIDDING PHASE HELPERS ===

def active_seat_bidding(dealer: str, bid_count: int) -> str:
    """
    Get the active seat during bidding.

    Args:
        dealer: Dealer's seat string
        bid_count: Number of bids made so far

    Returns:
        Active seat string
    """
    dealer_idx = seat_index(dealer)
    return seat_from_index(dealer_idx + bid_count)


def active_seat_play(trick_leader: str, cards_played: int) -> str:
    """
    Get the active seat during play.

    Args:
        trick_leader: Trick leader's seat string
        cards_played: Number of cards played in current trick (0-3)

    Returns:
        Active seat string
    """
    leader_idx = seat_index(trick_leader)
    return seat_from_index(leader_idx + cards_played)

File: frontend/src/utils/seats.js
/**
 * Seat utilities for bridge position calculations.
 *
 * This module provides a single source of truth for all seat-related
 * calculations across the frontend. Use these functions instead of
 * inline seat math to ensure consistency.
 *
 * Seat Mapping:
 *   NORTH = 0, EAST = 1, SOUTH = 2, WEST = 3
 *
 * The modulo-4 clock:
 *   - Partner is always +2 (opposite)
 *   - LHO (Left Hand Opponent) is always +1 (clockwise)
 *   - RHO (Right Hand Opponent) is always +3 (or -1)
 *
 * Usage:
 *   import { partner, lho, rho, relativePosition, displayName } from '../utils/seats';
 *
 *   // Get partner of South
 *   partner('S')  // Returns 'N'
 *
 *   // Get display name relative to user
 *   displayName('E', 'S')  // Returns 'LHO'
 *
 *   // Check if two seats are partners
 *   isPartner('N', 'S')  // Returns true
 */

// === CONSTANTS ===

// Integer indices (for modulo arithmetic)
export const NORTH = 0;
export const EAST = 1;
export const SOUTH = 2;
export const WEST = 3;

// Partnerships
export const NS = 0; // North-South
export const EW = 1; // East-West

// String representations
export const SEATS = ['N', 'E', 'S', 'W'];
export const SEAT_NAMES = { N: 'North', E: 'East', S: 'South', W: 'West' };
export const RELATIVE_NAMES = ['You', 'LHO', 'Partner', 'RHO'];

// Partnership sets
export const NS_SIDE = new Set(['N', 'S']);
export const EW_SIDE = new Set(['E', 'W']);

// Direct lookup maps (for performance when not using indices)
export const PARTNERS = { N: 'S', S: 'N', E: 'W', W: 'E' };
export const NEXT_PLAYER = { N: 'E', E: 'S', S: 'W', W: 'N' }; // LHO / clockwise
export const PREV_PLAYER = { N: 'W', E: 'N', S: 'E', W: 'S' }; // RHO / counter-clockwise


// === NORMALIZATION ===

/**
 * Normalize a position to single-letter format.
 *
 * @param {string|null|undefined} position - Position string ('North', 'N', 'north', etc.)
 * @returns {string} Single letter format ('N', 'E', 'S', 'W')
 *
 * @example
 * normalize('North') // 'N'
 * normalize('s') // 'S'
 * normalize(null) // 'N'
 */
export function normalize(position) {
  if (!position) return 'N'; // Default

  const pos = String(position).trim().toUpperCase();

  // Already single letter
  if (SEATS.includes(pos)) {
    return pos;
  }

  // Full name
  const fullMap = { NORTH: 'N', EAST: 'E', SOUTH: 'S', WEST: 'W' };
  return fullMap[pos] || 'N';
}


/**
 * Convert seat string to index (0-3).
 *
 * @param {string} seat - Seat string ('N', 'E', 'S', 'W' or full name)
 * @returns {number} Integer index (0=N, 1=E, 2=S, 3=W)
 */
export function seatIndex(seat) {
  return SEATS.indexOf(normalize(seat));
}


/**
 * Convert index to seat string.
 *
 * @param {number} index - Integer index (0-3, will be wrapped with modulo)
 * @returns {string} Seat string ('N', 'E', 'S', 'W')
 */
export function seatFromIndex(index) {
  // Handle negative indices correctly with ((index % 4) + 4) % 4
  return SEATS[((index % 4) + 4) % 4];
}


// === RELATIONSHIP FUNCTIONS ===

/**
 * Get the partner of a seat.
 *
 * @param {string} seat - Seat string
 * @returns {string} Partner's seat string
 *
 * @example
 * partner('S') // 'N'
 * partner('E') // 'W'
 */
export function partner(seat) {
  const s = normalize(seat);
  return PARTNERS[s];
}


/**
 * Get the Left Hand Opponent (next player clockwise).
 *
 * @param {string} seat - Seat string
 * @returns {string} LHO's seat string
 *
 * @example
 * lho('S') // 'W'
 * lho('N') // 'E'
 */
export function lho(seat) {
  const s = normalize(seat);
  return NEXT_PLAYER[s];
}


/**
 * Get the Right Hand Opponent (previous player clockwise).
 *
 * @param {string} seat - Seat string
 * @returns {string} RHO's seat string
 *
 * @example
 * rho('S') // 'E'
 * rho('N') // 'W'
 */
export function rho(seat) {
  const s = normalize(seat);
  return PREV_PLAYER[s];
}


/**
 * Get the partnership of a seat.
 *
 * @param {string} seat - Seat string
 * @returns {number} Partnership index (0=NS, 1=EW)
 */
export function partnership(seat) {
  return seatIndex(seat) % 2;
}


/**
 * Get the partnership name of a seat.
 *
 * @param {string} seat - Seat string
 * @returns {string} Partnership string ('NS' or 'EW')
 */
export function partnershipStr(seat) {
  return partnership(seat) === NS ? 'NS' : 'EW';
}


/**
 * Check if two seats are partners.
 *
 * @param {string} seat1 - First seat string
 * @param {string} seat2 - Second seat string
 * @returns {boolean} True if seats are partners
 */
export function isPartner(seat1, seat2) {
  return partner(seat1) === normalize(seat2);
}


/**
 * Check if two seats are opponents.
 *
 * @param {string} seat1 - First seat string
 * @param {string} seat2 - Second seat string
 * @returns {boolean} True if seats are opponents
 */
export function isOpponent(seat1, seat2) {
  return partnership(seat1) !== partnership(seat2);
}


/**
 * Check if two seats are on the same side (partners).
 *
 * @param {string} seat1 - First seat string
 * @param {string} seat2 - Second seat string
 * @returns {boolean} True if seats are on the same partnership
 */
export function sameSide(seat1, seat2) {
  return partnership(seat1) === partnership(seat2);
}


// === RELATIVE POSITION ===

/**
 * Get the relative position of target from hero's perspective.
 *
 * Uses modulo-4 clock arithmetic:
 *   0 = Self
 *   1 = LHO (Left Hand Opponent)
 *   2 = Partner
 *   3 = RHO (Right Hand Opponent)
 *
 * @param {string} target - Target seat string
 * @param {string} hero - Hero/user seat string
 * @returns {number} Relative position index (0-3)
 *
 * @example
 * relativePosition('W', 'S') // 1 (LHO)
 * relativePosition('N', 'S') // 2 (Partner)
 * relativePosition('E', 'S') // 3 (RHO)
 */
export function relativePosition(target, hero) {
  const diff = seatIndex(target) - seatIndex(hero);
  return ((diff % 4) + 4) % 4; // Handle negative correctly
}


/**
 * Get a display name for a seat.
 *
 * @param {string} seat - Target seat string
 * @param {string} [hero='S'] - Hero/user seat string
 * @param {boolean} [relative=true] - If true, return relative name ('Partner');
 *                                    if false, return absolute name ('North')
 * @returns {string} Display name string
 *
 * @example
 * displayName('N', 'S') // 'Partner'
 * displayName('N', 'S', false) // 'North'
 * displayName('E', 'S') // 'RHO'
 */
export function displayName(seat, hero = 'S', relative = true) {
  const s = normalize(seat);
  if (relative) {
    const rel = relativePosition(s, hero);
    return RELATIVE_NAMES[rel];
  }
  return SEAT_NAMES[s];
}


/**
 * Get the role description for a bidder relative to the user.
 *
 * @param {string} bidder - Bidder's seat string
 * @param {string} user - User's seat string
 * @returns {string} Role string with seat name, e.g., 'North (Partner)'
 *
 * @example
 * bidderRole('N', 'S') // 'North (Partner)'
 * bidderRole('E', 'S') // 'East (Opponent)'
 * bidderRole('S', 'S') // 'South (You)'
 */
export function bidderRole(bidder, user) {
  const b = normalize(bidder);
  const u = normalize(user);

  const rel = relativePosition(b, u);
  const seatName = SEAT_NAMES[b];

  if (rel === 0) {
    return `${seatName} (You)`;
  } else if (rel === 2) {
    return `${seatName} (Partner)`;
  } else {
    return `${seatName} (Opponent)`;
  }
}


// === PLAY PHASE HELPERS ===

/**
 * Get the dummy seat given the declarer.
 *
 * @param {string} declarer - Declarer's seat string
 * @returns {string} Dummy's seat string (declarer's partner)
 */
export function dummy(declarer) {
  return partner(declarer);
}


/**
 * Get the opening leader seat given the declarer.
 *
 * @param {string} declarer - Declarer's seat string
 * @returns {string} Opening leader's seat string (LHO of declarer)
 */
export function openingLeader(declarer) {
  return lho(declarer);
}


/**
 * Check if a seat is on the declaring side.
 *
 * @param {string} seat - Seat to check
 * @param {string} declarer - Declarer's seat string
 * @returns {boolean} True if seat is declarer or dummy
 */
export function isDeclaringSide(seat, declarer) {
  return sameSide(seat, declarer);
}


/**
 * Check if a seat is on the defending side.
 *
 * @param {string} seat - Seat to check
 * @param {string} declarer - Declarer's seat string
 * @returns {boolean} True if seat is a defender
 */
export function isDefendingSide(seat, declarer) {
  return !isDeclaringSide(seat, declarer);
}


// === BIDDING PHASE HELPERS ===

/**
 * Get the active seat during bidding.
 *
 * @param {string} dealer - Dealer's seat string
 * @param {number} bidCount - Number of bids made so far
 * @returns {string} Active seat string
 */
export function activeSeatBidding(dealer, bidCount) {
  const dealerIdx = seatIndex(dealer);
  return seatFromIndex(dealerIdx + bidCount);
}


/**
 * Get the active seat during play.
 *
 * @param {string} trickLeader - Trick leader's seat string
 * @param {number} cardsPlayed - Number of cards played in current trick (0-3)
 * @returns {string} Active seat string
 */
export function activeSeatPlay(trickLeader, cardsPlayed) {
  const leaderIdx = seatIndex(trickLeader);
  return seatFromIndex(leaderIdx + cardsPlayed);
}


/**
 * Determine success from NS perspective.
 *
 * @param {boolean} nsIsDeclarer - Whether NS is the declaring side
 * @param {number} actualTricksNs - Actual tricks NS took
 * @param {number} requiredTricks - Tricks required to make/set contract
 * @returns {boolean} True if NS succeeded (made contract or set opponents)
 */
export function nsSuccess(nsIsDeclarer, actualTricksNs, requiredTricks) {
  if (nsIsDeclarer) {
    // NS declaring: need to make the contract
    return actualTricksNs >= requiredTricks;
  } else {
    // NS defending: need to take enough tricks to set the contract
    // If EW needs 'requiredTricks', NS needs (14 - requiredTricks) to set
    // Equivalently: NS succeeds if EW got fewer than requiredTricks
    // Since EW tricks = 13 - actualTricksNs, EW < required means:
    // 13 - actualTricksNs < requiredTricks
    // actualTricksNs > 13 - requiredTricks
    return actualTricksNs > (13 - requiredTricks);
  }
}
